editor/src/messages/portfolio/document/utility_types/network_interface.rs

```diff
@@ -453,114 +454,143 @@ impl NodeNetworkInterface {
-				return concrete!(());
-			};
-			let Some(node) = current_network.nodes.get(&node_id) else {
-				log::error!("Could not get node {node_id} in input_type");
-				return concrete!(());
-			};
-			let node_id_path = [network_path, &[node_id]].concat().clone();
-			match &node.implementation {
-				DocumentNodeImplementation::Network(nested_network) => {
-					let downstream_connection = nested_network
-						.nodes
-						.iter()
-						.flat_map(|(node_id, node)| node.inputs.iter().enumerate().map(|(input_index, input)| (InputConnector::node(*node_id, input_index), input)))
-						.chain(nested_network.exports.iter().enumerate().map(|(export_index, export)| (InputConnector::Export(export_index), export)))
-						.find(|(_, input)| {
-							if let NodeInput::Network { import_index, .. } = input {
-								*import_index == input_connector.input_index()
-							} else {
-								false
-							}
-						});
-					if let Some((input_connector, _)) = downstream_connection {
-						Some(self.input_type(&input_connector, &node_id_path))
-					}
-					// Nothing is connected to the import
-					else {
-						Some(concrete!(()))
+	/// Try and get the [`DocumentNodeDefinition`] for a node
+	pub fn get_node_definition(&self, network_path: &[NodeId], node_id: NodeId) -> Option<&DocumentNodeDefinition> {
+		let metadata = self.node_metadata(&node_id, network_path)?;
+		resolve_document_node_type(metadata.persistent_metadata.reference.as_ref()?)
+	}
+
+	/// Try and get the [`Type`] for any [`InputConnector`] based on the `self.resolved_types`.
+	fn node_type_from_compiled(&self, input_connector: &InputConnector, network_path: &[NodeId]) -> Option<(Type, TypeSource)> {
+		let (node_id, input_index) = match *input_connector {
+			InputConnector::Node { node_id, input_index } => (node_id, input_index),
+			InputConnector::Export(export_index) => {
+				let Some((encapsulating_node_id, encapsulating_node_id_path)) = network_path.split_last() else {
+					// The outermost network export defaults to an ArtboardGroup.
+					return Some((concrete!(graphene_core::ArtboardGroup), TypeSource::OuterMostExportDefault));
+				};
+
+				let output_type = self.output_types(encapsulating_node_id, encapsulating_node_id_path).into_iter().nth(export_index).flatten();
+				if output_type.is_none() {
+					warn!("Could not find output type for export node");
+				}
+				return output_type;
+			}
+		};
+		let Some(current_network) = self.network(network_path) else {
+			log::error!("Could not get current network in input_type");
+			return None;
+		};
+		let Some(node) = current_network.nodes.get(&node_id) else {
+			log::error!("Could not get node {node_id} in input_type");
+			return None;
+		};
+		// If the input_connector is a NodeInput::Value, return the type of the tagged value.
+		if let Some(value) = node.inputs.get(input_index).and_then(|input| input.as_value()) {
+			return Some((value.ty(), TypeSource::TaggedValue));
```

Comment:
Im not sure if the type of the tagged value should directly be returned. It should be in the resolved types.

Reply:
It may be in the resolved types (unless the node is disconnected). However it will always resolve to the same type so we may as well just return here.

Reply:
I believe that the plan for disconnected node groups is to execute them seperately in order to resolve all their types/thumbnails